In [5]:
#Q3 part 1: Spelling checker with LED distance
import sys
import re
from collections import Counter

In [4]:
# Bottom-Up Levenshtein Distance function
def levenshtein_distance(word1, word2):
    m, n = len(word1), len(word2)
    dp = [[0]*(n+1) for _ in range(m+1)]

    for i in range(m+1):
        dp[i][0] = i
    for j in range(n+1):
        dp[0][j] = j

    for i in range(1, m+1):
        for j in range(1, n+1):
            if word1[i-1] == word2[j-1]:
                cost = 0
            else:
                cost = 1
            dp[i][j] = min(dp[i-1][j] + 1,    # deletion
                           dp[i][j-1] + 1,    # insertion
                           dp[i-1][j-1] + cost)  # substitution
    return dp[m][n]

# Load dictionary words and frequencies
def load_dictionary(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        text = f.read().lower()
    words = re.findall(r'\b[a-z]+\b', text)
    return Counter(words)

# Load misspelled words
def load_misspelled(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        text = f.read().lower()
    return re.findall(r'\b[a-z]+\b', text)

# Find suggestions for a misspelled word
def get_suggestions(word, dictionary, max_edit_distance, min_word_length):
    suggestions = []
    for dict_word in dictionary.keys():
        if len(dict_word) >= min_word_length:
            distance = levenshtein_distance(word, dict_word)
            if distance <= max_edit_distance:
                suggestions.append((dict_word, (distance, dictionary[dict_word])))
    suggestions.sort(key=lambda x: (x[1][0], -x[1][1]))  # sort by (edit distance, frequency descending)
    return suggestions

# Main function to run in Jupyter
def run_spell_checker(max_suggestions=7, dictionary_file='dictionary.txt', misspelled_file='misspelled.txt'):
    dictionary = load_dictionary(dictionary_file)
    print(f"Dictionary size: {len(dictionary)}")

    misspelled_words = load_misspelled(misspelled_file)

    max_edit_distance = 2
    min_word_length = 3

    for word in misspelled_words:
        if word not in dictionary:
            suggestions = get_suggestions(word, dictionary, max_edit_distance, min_word_length)
            suggestions = suggestions[:max_suggestions]
            suggestion_dict = {w: info for w, info in suggestions}
            print(f"\n- Suggestions for '{word}': {suggestion_dict}")

# Example Usage in Jupyter:
# run_spell_checker(7)

In [6]:
run_spell_checker(7)


Dictionary size: 29056

- Suggestions for 'homan': {'woman': (1, 323), 'human': (1, 170), 'coman': (1, 17), 'roman': (1, 8), 'man': (2, 1648), 'women': (2, 385), 'home': (2, 294)}

- Suggestions for 'spote': {'spoke': (1, 218), 'spite': (1, 117), 'spot': (1, 76), 'spots': (1, 12), 'smote': (1, 4), 'spore': (1, 3), 'some': (2, 1536)}

- Suggestions for 'belst': {'best': (1, 268), 'beast': (1, 26), 'belt': (1, 11), 'felt': (2, 697), 'west': (2, 280), 'rest': (2, 206), 'else': (2, 201)}

- Suggestions for 'effrts': {'efforts': (1, 103), 'effort': (2, 130), 'effects': (2, 82), 'forts': (2, 8), 'exerts': (2, 3), 'effete': (2, 1)}

- Suggestions for 'speling': {'spelling': (1, 4), 'feeling': (2, 362), 'seeing': (2, 207), 'speaking': (2, 185), 'swelling': (2, 164), 'smiling': (2, 161), 'opening': (2, 146)}

- Suggestions for 'perrfect': {'perfect': (1, 39), 'prefect': (2, 2)}

- Suggestions for 'avorage': {'average': (1, 18), 'voyage': (2, 12), 'forage': (2, 7), 'storage': (2, 3), 'averages':

Markdown Outputs: Dictionary size: 29056

- Suggestions for 'homan': {'woman': (1, 323), 'human': (1, 170), 'coman': (1, 17), 'roman': (1, 8), 'man': (2, 1648), 'women': (2, 385), 'home': (2, 294)}

- Suggestions for 'spote': {'spoke': (1, 218), 'spite': (1, 117), 'spot': (1, 76), 'spots': (1, 12), 'smote': (1, 4), 'spore': (1, 3), 'some': (2, 1536)}

- Suggestions for 'belst': {'best': (1, 268), 'beast': (1, 26), 'belt': (1, 11), 'felt': (2, 697), 'west': (2, 280), 'rest': (2, 206), 'else': (2, 201)}

- Suggestions for 'effrts': {'efforts': (1, 103), 'effort': (2, 130), 'effects': (2, 82), 'forts': (2, 8), 'exerts': (2, 3), 'effete': (2, 1)}

- Suggestions for 'speling': {'spelling': (1, 4), 'feeling': (2, 362), 'seeing': (2, 207), 'speaking': (2, 185), 'swelling': (2, 164), 'smiling': (2, 161), 'opening': (2, 146)}

- Suggestions for 'perrfect': {'perfect': (1, 39), 'prefect': (2, 2)}

- Suggestions for 'avorage': {'average': (1, 18), 'voyage': (2, 12), 'forage': (2, 7), 'storage': (2, 3), 'averages': (2, 1), 'averaged': (2, 1)}

- Suggestions for 'typo': {'type': (1, 84), 'two': (2, 1137), 'too': (2, 548), 'top': (2, 42), 'types': (2, 33), 'tips': (2, 16), 'tap': (2, 10)}

- Suggestions for 'misspeling': {}

- Suggestions for 'mlch': {'much': (1, 671), 'such': (2, 1436), 'each': (2, 411), 'march': (2, 134), 'rich': (2, 92), 'match': (2, 41), 'mack': (2, 22)}

- Suggestions for 'mistkes': {'mistakes': (1, 15), 'mistaken': (2, 59), 'mistake': (2, 39), 'mistress': (2, 24), 'mister': (2, 2), 'misses': (2, 2), 'mists': (2, 1)}

- Suggestions for 'hauunt': {'haunt': (1, 2), 'aunt': (2, 52), 'hunt': (2, 31), 'gaunt': (2, 5), 'taunt': (2, 1)}

- Suggestions for 'odoor': {'door': (1, 498), 'odour': (1, 19), 'odor': (1, 6), 'poor': (2, 128), 'floor': (2, 105), 'doors': (2, 47), 'doom': (2, 6)}

- Suggestions for 'colone': {'colonel': (1, 143), 'colony': (1, 57), 'colon': (1, 9), 'cologne': (1, 3), 'alone': (2, 337), 'close': (2, 219), 'colonies': (2, 202)}

